src/libraries/System.IO/tests/MemoryStream/MemoryStreamTests.cs

```diff
@@ -82,6 +84,34 @@ public static void MemoryStream_WriteToTests_Negative()
             }
         }
 
+        public static IEnumerable<object[]> MemoryStream_PositionOverflow_Throws_MemberData() =>
+            from mode in Enum.GetValues<SeekMode>()
+            from bufferContext in
+                new (int bufferSize, int origin)[]
+                {
+                    (0, 0),
+                    (1, 0),
+                    (1, 1),
+                    (10, 0),
+                    (10, 5),
+                    (10, 10),
+                    (Array.MaxLength, 0),
+                    (Array.MaxLength, Array.MaxLength)
+                }
+            select new object[] {mode, bufferContext.bufferSize, bufferContext.origin};
+
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.Is64BitProcess))]
+        [MemberData(nameof(MemoryStream_PositionOverflow_Throws_MemberData))]
+        public void MemoryStream_SeekOverflow_Throws(SeekMode mode, int bufferSize, int origin)
+        {
+            byte[] buffer = new byte[bufferSize];
+            using (MemoryStream ms = new MemoryStream(buffer, origin, buffer.Length - origin, true))
+            {
+                Seek(mode, ms, int.MaxValue - origin);
+                Assert.Throws<ArgumentOutOfRangeException>(() => Seek(mode, ms, (long)int.MaxValue - origin + 1));
```

Comment:
Can you please test with an offset of `long.MinValue + 1`. The check we discussed in https://github.com/dotnet/runtime/pull/88572/files/6880c4d6d148f46a878f929303fd2472e8fbc363#r1297747637 may be there to prevent an overflow after the `(int)` conversion.

Reply:
Yes, removing the `unchecked(loc + offset) < _origin` check was a mistake.  Because `(int)offset` used for calculating `tempPosition` will just cut the upper 32 bit from the long value, turning certain negative offset values into positive `tempPosition` values that are larger than `_origin`, thus not correctly failing on offset values representing invalid negative positions.

Reply:
I have added two Seek calls with `long.MinValue + 1` and `long.MaxValue - 1` as offset and checked that an exception is correctly raised in all cases.
Indeed I have manually checked that these calls are not always sending an exception when the `unchecked(loc + offset) < _origin` check is not present.